File: leader_arm_driver/src/leader_joint_publisher.cpp
#include <ros/ros.h>
#include <sensor_msgs/JointState.h>
#include <vector>
#include <string>
#include <cmath>

class ManipulatorJointPublisher
{
public:
    ManipulatorJointPublisher()
        : rate_(50.0), angle_(0.0)
    {
        joint_pub_ = nh_.advertise<sensor_msgs::JointState>("/joint_states", 10);

        // Must match your URDF exactly (case-sensitive!)
        joint_names_ = {
            "Joint01", "Joint02", "Joint03",
            "Joint04", "Joint05", "Joint06"
        };
    }

    void run()
    {
        while (ros::ok())
        {
            sensor_msgs::JointState js;
            js.header.stamp = ros::Time::now();
            js.name = joint_names_;

            // Simulate one moving joint and fixed values for others
            js.position = {
                std::sin(angle_),
                0.5,
                -0.3,
                0.2,
                -0.1,
                0.0
            };

            joint_pub_.publish(js);

            angle_ += 0.05;
            rate_.sleep();
        }
    }

private:
    ros::NodeHandle nh_;
    ros::Publisher joint_pub_;
    ros::Rate rate_;
    double angle_;
    std::vector<std::string> joint_names_;
};

int main(int argc, char** argv)
{
    ros::init(argc, argv, "manipulator_joint_publisher");
    ManipulatorJointPublisher node;
    node.run();
    return 0;
}

File: ros_controller/scripts/leader_joint_publisher.py
#!/usr/bin/env python
import rospy
from sensor_msgs.msg import JointState
import math
import time

class ManipulatorJointPublisher:
    def __init__(self):
        rospy.init_node('manipulator_joint_publisher')
        self.pub = rospy.Publisher('/joint_states', JointState, queue_size=10)
        self.rate = rospy.Rate(50)  # 50 Hz

        # Joint names must match your URDF
        self.joint_names = [
            'Joint01', 'Joint02', 'Joint03',
            'Joint04', 'Joint05', 'Joint06'
        ]

        # Initialize joint positions
        self.angle = 0.0

    def run(self):
        while not rospy.is_shutdown():
            js = JointState()
            js.header.stamp = rospy.Time.now()
            js.name = self.joint_names

            # Example: oscillate the first joint, keep others fixed
            js.position = [
                math.sin(self.angle),
                0.5,
                -0.3,
                0.2,
                -0.1,
                0.0
            ]

            self.pub.publish(js)

            self.angle += 0.05
            self.rate.sleep()

if __name__ == '__main__':
    try:
        node = ManipulatorJointPublisher()
        node.run()
    except rospy.ROSInterruptException:
        pass
